### PR TITLE
feat: Node action button group on MachineDetails MAASENG-1223

### PR DESCRIPTION
**Pull request by ndv99**
## Done

- Added group of contextual menus and buttons to replace "Take action" dropdown
- Implemented on machine details
- Hide group and replace with "Menu dropdown" on smaller screens

### MAAS deployment

To run this branch you will need access to one of the following MAAS deployments:

- [Karura](/HACKING.md#karura)
- [Bolla](/HACKING.md#bolla)
- [A development MAAS](/HACKING.md#development-deployment)
- [A local snap MAAS](/HACKING.md#snap-deployment) (this will not usually have machines)

### Running the branch

You can run this branch by:

- Serving with [dotrun](/HACKING.md#maas-ui-development-setup)
- [Building in a development MAAS](/HACKING.md#running-maas-ui-from-a-development-maas)

### QA steps

- Go to the details page for a machine
- Ensure the action group renders correctly
- Click on each available action
- Ensure it opens the correct form
- Shrink your window horizontally
- Ensure the group is replaced by a "Menu" dropdown wrapping occurs within the button group
- Ensure that each available action in the menu opens the correct form

## Fixes

Fixes https://warthogs.atlassian.net/browse/MAASENG-1223

### Before
image

### After
image

#### With lock switch
image

#### Small screen
image

### CODE REVIEW

**Comment by petermakowski** [comment]
@ndv99 Just had a quick look, this is looking really good!

Just a small thing I noticed: seems like the height of each button in the new button group should be the same as the height of existing buttons on the machine list page (e.g. "Add hardware"), but they're much smaller now.

**Comment by ndv99** [comment]
> @ndv99 Just had a quick look, this is looking really good!
> 
> Just a small thing I noticed: seems like the height of each button in the new button group should be the same as the height of existing buttons on the machine list page (e.g. "Add hardware"), but they're much smaller now.

I'll ask @amylily1011 about the implementation of that. For this PR I've just kept it on the MachineDetails page only

**Comment by amylily1011** [comment]
Yup @petermakowski is right, I think the button looks quite small. There seems to be more padding for each button in the button set via the design as well. Does the figma file provide you with any padding and sizes? @ndv99 I can find some references for the size.

**Comment by ndv99** [comment]
> Yup @ petermakowski is right, I think the button looks quite small. There seems to be more padding for each button in the button set via the design as well. Does the figma file provide you with any padding and sizes? @ndv99 I can find some references for the size.

If you could actually that'd be amazing, there's a lot of wireframes in the figma file and I'm not sure which is correct, I've just been working based on the image in the spec

File: src/app/machines/views/MachineDetails/MachineHeader/index.scss [hunk 1]
```diff
@@ -0,0 +1,19 @@
+@mixin MachineHeader {
+  .p-table-menu {
+    margin-right: $spv--large;
+  }
+
+  // These media queries are for when the screen is too small to display the action group without wrapping.
+  // Wrapping begins before $breakpoint-small, hence the direct reference to pixel width.
+  @media screen and (max-width: 792px) {
```
Comment (petermakowski) [inline src/app/machines/views/MachineDetails/MachineHeader/index.scss:8]: Could you please add this as a new custom breakpoint to `_settings.scss` which will help us keep track of them all in a single place. Ideally using a single breakpoint and `$breakpoint-name + 1` `$breakpoint-name - 1` convention that we use in vanilla.

File: src/app/base/components/NodeActionMenuGroup/NodeActionMenuGroup.test.tsx [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+        showCount
+      />
+    );
+
+    await openMenu(Labels.Actions);
+    const actionsMenu = getSubMenu(Labels.Actions);
+
+    expect(
+      within(getActionCount(actionsMenu, NodeActions.COMMISSION)).getByText("3")
+    );
```
[the remark this answers is not in the document]
Comment (ndv99) [reply src/app/base/components/NodeActionMenuGroup/NodeActionMenuGroup.test.tsx:205]: I'm actually surprised as well - a lot of the tests were copied and modified from NodeActionMenu, that must have slipped past. I'll rectify it here and in NodeActionMenu

File: src/app/machines/views/MachineDetails/MachineHeader/index.scss [hunk 3]
```diff
@@ -0,0 +1,19 @@
+@mixin MachineHeader {
+  .p-table-menu {
+    margin-right: $spv--large;
+  }
```
[the remark this answers is not in the document]
Comment (ndv99) [reply src/app/machines/views/MachineDetails/MachineHeader/index.scss:4]: I didn't know that existed actually, I'll try it out

File: src/app/machines/views/MachineDetails/MachineHeader/index.scss (hunk 1 above)
Comment (ndv99) [reply src/app/machines/views/MachineDetails/MachineHeader/index.scss:8]: Sure, although not sure what to name it - it's closer to `small` than `large` so I don't feel like `medium` would fit, and `kinda-small` just doesn't seem right lol

File: src/app/base/components/SectionHeader/SectionHeader.tsx [hunk 4]
```diff
@@ -125,6 +127,7 @@ const SectionHeader = <P,>({
         size={headerSize}
         title={sidePanelTitle}
       />
+      {actionMenuGroup && <>{actionMenuGroup}</>}
```
[the remark this answers is not in the document]
Comment (ndv99) [reply src/app/base/components/SectionHeader/SectionHeader.tsx:130]: Sure, I'll commit that

File: src/app/base/components/NodeActionMenuGroup/index.scss [hunk 5]
```diff
@@ -0,0 +1,26 @@
+@mixin NodeActionMenuGroup {
+  .p-button-group {
+    display: inline-flex;
+    flex-wrap: nowrap;
+    border: 1px solid #757575;
```
[the remark this answers is not in the document]
Comment (ndv99) [reply src/app/base/components/NodeActionMenuGroup/index.scss:5]: Sure thing, I'll double check to be sure

File: src/app/base/components/NodeActionMenuGroup/index.scss [hunk 6]
```diff
@@ -0,0 +1,26 @@
+@mixin NodeActionMenuGroup {
+  .p-button-group {
```
[the remark this answers is not in the document]
Comment (ndv99) [reply src/app/base/components/NodeActionMenuGroup/index.scss:2]: I'll do the rename I think

File: src/app/base/components/NodeActionMenuGroup/index.scss (hunk 5 above)
[the remark this answers is not in the document]
Comment (ndv99) [reply src/app/base/components/NodeActionMenuGroup/index.scss:5]: That colour is too light, I think the correct one is `$colors--light-theme--border-high-contrast`

File: src/app/machines/views/MachineDetails/MachineHeader/index.scss (hunk 3 above)
[the remark this answers is not in the document]
Comment (ndv99) [reply src/app/machines/views/MachineDetails/MachineHeader/index.scss:4]: That works. Will commit

File: src/app/machines/views/MachineDetails/MachineHeader/index.scss (hunk 1 above)
Comment (ndv99) [reply src/app/machines/views/MachineDetails/MachineHeader/index.scss:8]: I think I'll settle on `medium-small`

**Pull request merged**